### ISSUE TITLE
Can Lint/AssignmentInCondition trigger in more cases?

### ISSUE THREAD

**Issue by wollkind** [rubocop-hq/rubocop#4446]
I see that this came up once before in #3323 but the filing user closed it on his own.  Sorry if this is an unwelcome duplicate.

I was just surprised to discover that this cop only triggers inside if/when/unless conditionals.  This came up because I had a bug in my code with an accidental assignment inside an array select block where I meant to check for equality.

`oops = array.select { |element| element.property = blah }`

After I discovered this, I was surprised that rubocop wasn't warning me, as this is exactly the kind of thing I'd like it to be alert for.  When I finally found the line, I saw that my IDE (Rubymine) was flagging it, but Rubymine doesn't automatically report on/inspect your entire codebase, so I didn't notice the warning.

I don't know if it's hard/undesirable/both/neither to expand this cop to look at all conditionals, but if that were an option, I'd surely turn it on.

We are using Rubocop 0.48.1 but I don't see any sign that this has changed in the latest version.

**Comment by backus** [comment]
This seems like a good idea! Using [`rdl`'s](https://github.com/plum-umd/rdl) `rdl_query` method, I pulled up a list of other methods in the stdlib that we could check:

```sh
$ rdl_query '(...) { (...) -> %bool } -> .'
Array#count: () {(t) -> false or true} -> Bignum
Array#delete_if: () {(t) -> false or true} -> Array<t>
Array#drop_while: () {(t) -> false or true} -> Array<t>
Array#index: () {(t) -> false or true} -> Bignum
Array#keep_if: () {(t) -> false or true} -> Array<t>
Array#reject: () {(t) -> false or true} -> Array<t>
Array#reject!: () {(t) -> false or true} -> Array<t>
Array#rindex: () {(t) -> false or true} -> Bignum
Array#select: () {(t) -> false or true} -> Array<t>
Array#select!: () {(t) -> false or true} -> Array<t>
Array#take_while: () {(t) -> false or true} -> Array<t>
Enumerable#all?: () {(t) -> false or true} -> false or true
Enumerable#any?: () {(t) -> false or true} -> false or true
Enumerable#count: () {(t) -> false or true} -> Fixnum
Enumerable#detect: (?Proc ifnone) {(t) -> false or true} -> t or nil
Enumerable#drop_while: () {(t) -> false or true} -> Array<t>
Enumerable#find_all: () {(t) -> false or true} -> Array<t>
Enumerable#find_index: () {(t) -> false or true} -> Fixnum or nil
Enumerable#none?: () {(t) -> false or true} -> false or true
Enumerable#one?: () {(t) -> false or true} -> false or true
Enumerable#partition: () {(t) -> false or true} -> [Array<t>, Array<t>]
Enumerable#reject: () {(t) -> false or true} -> Array<t>
Enumerable#take_while: () {(t) -> false or true} -> Array<t>
Hash#delete_if: () {(k, v) -> false or true} -> Hash<k, v>
Hash#keep_if: () {(k, v) -> false or true} -> Hash<k, v>
Hash#reject: () {(k, v) -> false or true} -> Hash<k, v>
Hash#reject!: () {(k, v) -> false or true} -> Hash<k, v>
Hash#select: () {(k, v) -> false or true} -> Hash<k, v>
Hash#select!: () {(k, v) -> false or true} -> Hash<k, v>
Range#bsearch: () {(t) -> false or true} -> u or nil
Set#delete_if: () {(t) -> false or true} -> self
Set#keep_if: () {(t) -> false or true} -> self
Set#reject!: () {(t) -> false or true} -> self or nil
Set#select!: () {(t) -> false or true} -> self or nil
```

So qualifying methods would be:

- `#all?`
- `#any?`
- `#bsearch`
- `#count`
- `#delete_if`
- `#detect`
- `#drop_while`
- `#find_all`
- `#find_index`
- `#index`
- `#keep_if`
- `#none?`
- `#one?`
- `#partition`
- `#reject`
- `#reject!`
- `#rindex`
- `#select`
- `#select!`
- `#take_while`

**Comment by bbatsov** [comment]
Nice!

**Comment by Drenmi** [comment]
Cool stuff! Should this be a separate cop, though? 🤔
